src/ipc/handler.c

```diff
@@ -70,6 +70,56 @@
 #define iGS(x) ((x >> SOF_GLB_TYPE_SHIFT) & 0xf)
 #define iCS(x) ((x >> SOF_CMD_TYPE_SHIFT) & 0xfff)
 
+/*
+ * IPC ABI version compatibility rules :-
+ *
+ * 1) FW binaries will only support one MAJOR ABI version which is advertised
+ *    to host at FW boot.
+ *
+ * 2) Host drivers will support the current and older MAJOR ABI versions of
+ *    the IPC ABI (up to a certain age to be determined by market information).
+ *
+ * 3) MINOR and PATCH ABI versions can differ between host and FW but must be
+ *    backwards compatible on both host and FW.
+ *
+ *    IPC messages sizes can be different for sender and receiver if MINOR or
+ *    PATCH ABI versions differ as new fields can be added to the end of
+ *    messages.
+ *
+ *    i) Sender > receiver: receiver only copies it's own ABI structure size.
+ *
+ *    ii) Receiver > sender: receiver copies its's own ABI size and zero pads
+ *                           new fields. i.e. new structure fields must be non
+ *                           zero to be activated.
+ *
+ *    Guidelines for extending ABI compatible messages :-
+ *
+ *    i) Use reserved fields.
+ *    ii) Grow structure at the end.
+ *    iii) Iff (i) and (ii) are not possible then MAJOR ABI is bumped.
+ */
+
+#define _IPC_COPY_CMD(rx, tx, rx_size)					\
+	do {								\
+		if (rx_size > tx->size) {				\
+			memcpy(rx, tx, tx->size);			\
+			bzero((void *)rx + tx->size, rx_size - tx->size);\
+			trace_ipc("ipc: hdr 0x%x rx (%d) > tx (%d)",	\
+				  rx->cmd, rx_size, tx->size);		\
+		} else if (tx->size > rx_size) {			\
+			memcpy(rx, tx, rx_size);			\
+			trace_ipc("ipc: hdr 0x%x tx (%d) > rx (%d)",	\
+				  rx->cmd, tx->size, rx_size);		\
+		} else							\
+			memcpy(rx, tx, rx_size);			\
+	} while (0)
+
+/* copies whole message from Tx to Rx, follows above ABI rules */
+#define IPC_COPY_CMD(rx, tx) \
+	_IPC_COPY_CMD(((struct sof_ipc_cmd_hdr *)&rx),			\
+			((struct sof_ipc_cmd_hdr *)tx),			\
+			sizeof(rx))
```

Comment:
If using buffer for ipc data handling
```
((struct sof_ipc_cmd_hdr *)tx)->size)
```
can copy data in full and if using buffer for ipc handling we wouldn't copy to smaller structs since allocation would look something like this:
```
struct sof_ipc_ctrl_data *data = rmalock(RZONE_IPC_BUFFER, SOF_MEM_CAPS_RAM, 
	((struct sof_ipc_cmd_hdr *)_ipc->comp_data)->size);
```
We then wouldn't have to relay on sending size down the line.